frame/evm/precompile/contracts/kton/src/lib.rs

```diff
@@ -108,14 +110,39 @@ impl<T: Config> Precompile for Kton<T> {
 				Ok((ExitSucceed::Returned, vec![], 20000))
 			}
 			Action::Withdraw(wd) => {
-				let withdraw_value = wd.kton_value.saturating_mul(helper);
-				T::KtonCurrency::transfer(
-					&con_caller,
+				// Ensure wkton is a contract
+				ensure!(
+					!crate::Module::<T>::is_contract_code_empty(&context.caller),
+					ExitError::Other("The caller must be wkton contract!".into())
+				);
+				// Ensure context's apparent_value is zero
+				ensure!(
+					context.apparent_value == U256::zero(),
+					ExitError::Other("The value in tx must be zero!".into())
+				);
+				// Ensure caller's balance is enough
+				let caller_kton = T::KtonAccountBasic::account_basic(&context.caller);
+				ensure!(caller_kton.balance >= wd.kton_value, ExitError::OutOfFund);
```

Comment:
`OutOfFund` - we should use the same error name for the insufficient balance situation. I didn't find any other name like this by searching this repo.  
BTW we should group errors in one file/crate i think

Reply:
The return error and `execute` function signature defined in the EVM crate. Thus you can not find it in this repo.

Reply:
> `OutOfFund` - we should use the same error name for the insufficient balance situation. I didn't find any other name like this by searching this repo.
> BTW we should group errors in one file/crate i think

No you can't.

Each error belong to its `mod pallets;`

Reply:
Yes, but we should have a name convention for error, event names, etc